// connection/connection.js

//WebSocket汎用サーバー
//extendsして使う
const ws  = require ('ws').Server;
//const EventEmitter = require('events').EventEmitter;
//全てのデータをJSON形式でやり取りをすることを推奨している。
class Connection{

    

    constructor(data){

        console.log("server working");
        this.ws = new ws(data);

        this.ws.on ('connection', (client,req) =>{this.onConnect(this,client,req)}); 
    }
    onConnect(self,client,req){

            //最初の接続時
            var id = req.headers['sec-websocket-key'];//ユニークなID
            self.onOpen(id,client,req);
            console.log("connect : "+id);
            client.on ('message',   (message) => {  self.onMessage(id,client,message) });
            client.on ('close',     ()        => { self.onClose(id,client,req)});
            client.on ('error' ,(e)  => {} );
        
        
    }
    //this.server.emit('close',id,req.connection.remoteAddress,client);
    onOpen(id,client,req){

    }
    onMessage(id,client,message){

    }
    onClose(id,client,req){
        
    }
    send(client,message){
       client.send( message );
    }

    broadcast (data) {
        this.ws.clients.forEach(
            function each(c) {
            if (c.readyState === c.OPEN) {
                 c.send (data);
             }
        });
    };
    
}

module.exports = Connection;
