// gwss/index.js
const gwss = require('./gwss');
//=========================================================================
//gwss用データフォーマット
//
//階層構造で、{type array{id data}} の形になる
//typeで含まれるデータを判断
//arrayをパースし、idがデータの主キー
//dataがデータ本体
//パーサーは用意しない
//data追加は instance.array[id] = data で行う。
//=========================================================================
class AssocArrayJsonFormat{
	//初期化 タイプ(とフォーマット)を決める
    constructor(type/*,jsonformat*/){
    	this.array = [];
    	this.type = type;
    	//this.format = jsonformat;
    }
    //=========================================================================
    //=========================================================================
   
    addId(array){
        var ret = [];
        Object.keys(array).forEach( (element) => {
            ret.push({id: element ,data: array[element]});
        });
        return ret;
    }
    //=========================================================================
    //=========================================================================
    //追加
    addData(id,data){
    	this.array[id] = data;
    }
    //削除
    deleteData(id){
    	delete this.array[id];
    }
    //typeを付加したJSONをstringで返す
    //送信用
    getAAJFString(){
    	var json = {type: this.type, array: this.addId(this.array)};
    	return JSON.stringify(json);
    }
    length(){
    	return Object.keys(this.array).length;
    }
    
}
gwss.AAJF = AssocArrayJsonFormat;
module.exports = gwss;
